File: src/146_LRU_缓存.cpp
// 146_LRU_缓存
// https://leetcode.cn/problems/lru-cache/description/?envType=study-plan-v2&envId=top-100-liked

// Below is the equivalent implementation of the given Rust LRUCache in C++.
// This implementation uses `std::unordered_map` to store key - value pairs
// and `std::list` to maintain the keys' order according to their usage,
// which acts as a queue to determine the least recently used key when eviction is needed.
#include <unordered_map>
#include <deque>
#include <list>

// If you specifically want to use a std::deque (double-ended queue) in C++ to maintain the order of keys,
// keep in mind that this structure does not support efficient random access deletion
// which is required to move an item to the end or front upon access.
// Due to this limitation, the time complexity for the mark_as_recently_used operation would be O(n),
// making the overall LRU cache less efficient.
// However, to provide a one-to-one translation using std::deque, here's how you can implement it:
class LRUCache
{
private:
    int capacity;
    std::unordered_map<int, int> map;
    std::deque<int> order;

    void mark_as_recently_used(int key)
    {
        // Remove key if it exists in the deque
        for (auto it = order.begin(); it != order.end(); ++it)
        {
            if (*it == key)
            {
                order.erase(it);
                break;
            }
        }
        // Add the key to the back to mark it as recently used
        order.push_back(key);
    }

public:
    LRUCache(int capacity) : capacity(capacity) {}

    int get(int key)
    {
        if (map.find(key) != map.end())
        {
            mark_as_recently_used(key);
            return map[key];
        }
        else
        {
            return -1;
        }
    }

    void put(int key, int value)
    {
        if (map.find(key) != map.end())
        {
            // Update the value
            map[key] = value;
            mark_as_recently_used(key);
        }
        else
        {
            if (map.size() == capacity)
            {
                // Evict the least recently used key
                int lru = order.front();
                map.erase(lru);
                order.pop_front();
            }
            map[key] = value;
            order.push_back(key);
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

#include <unordered_map>
#include <deque>
#include <list>

// In this C++ version:
using namespace std;

class LRUCache
{
private:
    int capacity;
    unordered_map<int, pair<int, std::list<int>::iterator>> cache;
    list<int> keys;

    void touch(int key)
    {
        auto &it = cache[key].second;
        keys.erase(it);
        keys.push_front(key);
        it = keys.begin();
    }

public:
    explicit LRUCache(int capacity) : capacity(capacity) {}

    int get(int key)
    {
        auto cacheIt = cache.find(key);
        if (cacheIt == cache.end())
        {
            return -1;
        }
        else
        {
            touch(key);
            return cacheIt->second.first;
        }
    }

    void put(int key, int value)
    {
        if (cache.find(key) != cache.end())
        {
            cache[key].first = value;
            touch(key);
        }
        else
        {
            if (cache.size() == capacity)
            {
                int lru = keys.back();
                cache.erase(lru);
                keys.pop_back();
            }
            keys.push_front(key);
            cache[key] = {value, keys.begin()};
        }
    }
};

// In this refactored version:

// capacity is explained as the maximum capacity of the cache.
// cache is a map that now has a more descriptive name and a comment explaining its use.
// keys is a list that maintains the order of keys based on their usage.
// The touch method is a new private method to encapsulate the logic of marking a key as recently used.
// This avoids duplication of code in get and put methods and makes them easier to read.
// Descriptive variable names and explicit variable types (auto replaced with explicit type where it may improve clarity).
// The explicit keyword is added to the constructor to prevent implicit conversions.
// Additional comments to explain the logic in the get and put methods.

File: src/146_LRU_缓存.py
# 146_LRU_缓存

from typing import Optional
from collections import OrderedDict
from collections import deque

# 通过但是超时


class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.maps = dict()
        self.order = deque()

    def get(self, key: int) -> int:
        # 超时瓶颈
        if key in self.maps:
            value = self.maps[key]
            self.mark_as_freq(key)
            return value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        # 超时瓶颈
        if key in self.maps:
            # Update the value for existing keys without checking the capacity
            self.maps[key] = value
            self.mark_as_freq(key)
        else:
            # Only check the capacity if it's a new key
            if len(self.maps) == self.capacity:
                # Remove the least recently used item
                old_key = self.order.pop()
                del self.maps[old_key]
            self.maps[key] = value
            self.order.appendleft(key)

    def mark_as_freq(self, key):
        # Move the key to the front (left side) of the deque to show it was recently accessed
        self.order.remove(key)
        self.order.appendleft(key)

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

# 使用Python OrderedDict()
# 有序字典


class LRUCache:
    def __init__(self, capacity: int):
        """ Initialize the LRU cache with a given capacity. """
        self.capacity = capacity
        self.od = OrderedDict()

    def get(self, key: int) -> int:
        """ Retrieve an item from the cache by key. Returns -1 if not found. """
        if key in self.od:
            self.od.move_to_end(key)
            return self.od[key]
        return -1

    def put(self, key: int, value: int):
        """ Put an item into the cache. If the cache exceeds its capacity, removes the least recently used item. """
        self.od[key] = value
        if len(self.od) > self.capacity:
            self.od.popitem(last=False)
        self.od.move_to_end(key)


cache = LRUCache(2)
cache.put(1, 1)
cache.put(2, 2)
print(cache.get(1))  # returns 1
cache.put(3, 3)  # L


# Python 双链表实现，比较复杂
class Node:
    __slots__ = 'prev', 'next', 'key', 'value'

    def __init__(self, key=0, value=0):
        self.key = key
        self.value = value


class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.dummy = Node()             # Sentinel node for ease of use
        self.dummy.prev = self.dummy
        self.dummy.next = self.dummy
        self.key_to_node = dict()

    def get_node(self, key: int) -> Optional[Node]:
        if key not in self.key_to_node:
            return None
        node = self.key_to_node[key]
        self.remove(node)
        self.push_front(node)
        return node

    def get(self, key: int) -> int:
        node = self.get_node(key)
        return node.value if node else -1

    def put(self, key: int, value: int) -> None:
        if key in self.key_to_node:
            node = self.get_node(key)
            node.value = value
        else:
            if len(self.key_to_node) == self.capacity:
                # Remove the least recently used node before adding a new one
                back_node = self.dummy.prev
                del self.key_to_node[back_node.key]
                self.remove(back_node)
            node = Node(key, value)
            self.key_to_node[key] = node
            self.push_front(node)

    def remove(self, node: Node) -> None:  # Corrected type hint
        node.prev.next = node.next
        node.next.prev = node.prev

    def push_front(self, node: Node) -> None:
        node.prev = self.dummy
        node.next = self.dummy.next
        self.dummy.next.prev = node
        self.dummy.next = node  # Corrected order of assignments
